### PR TITLE
Update IP recognizer regexes 

### LINKED ISSUE #1598: IPv6 is not getting completely anonymized
### Discussed in https://github.com/microsoft/presidio/discussions/1591

<div type='discussions-op-text'>

<sup>Originally posted by **Greatz08** May 15, 2025</sup>
image

I know it did hide initial part of ipv6 address, but i wanted it to completely hide that ipv6 address which didnt happened unfortunately as you can see in above image. So is the model not trained to hide complete ipv6 address ? or is it trained to behave like this only ?</div>

### PR DESCRIPTION
**Pull request by kennionblack**
## Change Description

This PR introduces new dedicated `IPv4_mapped` and `IPv4_embedded` patterns to reduce the complexity of the existing IPv4 and IPv6 regexes. These were previously included in the existing regexes and could in theory still be consolidated into one regex for each IP type, but this makes the regexes extremely long and difficult to debug.

This PR also addresses issues with the existing IPv6 regex. Because `\b` treats : as a non-word character, it fires at the position between :: and the following hex digits, allowing the `{1,7}:` segment to match only the prefix up to :: (e.g., `2400:c401::5054:ff:fe1b:b031` would match as just `2400:c401::`) even though the IP address continues after the ::. The new regex replaces `\b` with `(?<![\w:])` as a lookbehind and `(?![\w:.])` as a lookahead. The lookahead is what specifically prevents the truncation present in the original issue as it checks for a hex digit after :: and rejects the short match. 

Note that using `\w` instead of a hex-specific character class also prevents false positives from patterns in text that naturally include ::. For example, `std::cout` previously matched `d::c` as valid IPv6 because d and c are hex characters, but the t preceding them is a word character that the new lookbehind correctly rejects. The `\w` lookahead similarly ensures IPv6 addresses stapled to surrounding text (e.g., `fe80::1text`) are not matched, which is consistent with how the IPv4 pattern's `\b` boundaries already behave.

Finally, this PR adds an `IPv6_unspecified` pattern for the standalone :: address, which was previously just matching on `::`. `::` matched inside a number of non-IP contexts, including the C++ case given above. The new pattern uses the same lookahead/lookbehind approach used in the full IPv6 regex to only match :: when it is not adjacent to word characters.

Tests reuse the excellent test cases written by @navalev in #1635 with some minor modifications.

## Issue reference

Fixes #1598 

## Checklist

- [x] I have reviewed the [contribution guidelines](https://github.com/microsoft/presidio/blob/main/CONTRIBUTING.md)
- [x] I have signed the CLA (if required)
- [x] My code includes unit tests
- [x] All unit tests and lint checks pass locally
- [x] My PR contains documentation updates / additions if required

### CODE REVIEW

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py [hunk 1]
```diff
@@ -15,19 +15,29 @@ class IpRecognizer(PatternRecognizer):
...
         ),
         Pattern(
             "IPv6",
-            r"\b(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))\b",
+            r"(?<![\w:])(?:(?:[0-9A-Fa-f]{1,4}:){7}[0-9A-Fa-f]{1,4}|(?:[0-9A-Fa-f]{1,4}:){1,7}:|:(?::[0-9A-Fa-f]{1,4}){1,7}|(?:[0-9A-Fa-f]{1,4}:){1,6}:[0-9A-Fa-f]{1,4}|(?:[0-9A-Fa-f]{1,4}:){1,5}(?::[0-9A-Fa-f]{1,4}){1,2}|(?:[0-9A-Fa-f]{1,4}:){1,4}(?::[0-9A-Fa-f]{1,4}){1,3}|(?:[0-9A-Fa-f]{1,4}:){1,3}(?::[0-9A-Fa-f]{1,4}){1,4}|(?:[0-9A-Fa-f]{1,4}:){1,2}(?::[0-9A-Fa-f]{1,4}){1,5}|[0-9A-Fa-f]{1,4}:(?::[0-9A-Fa-f]{1,4}){1,6}|:(?::[0-9A-Fa-f]{1,4}){1,6})(?![\w:.])",
```
[the remark this answers is not in the document]
Comment (kennionblack) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py]: This was to prevent overlap with IPv4-embedded addresses. The best fix for this is likely to extend the lookahead to check for a digit after the period.

File: presidio-analyzer/tests/test_ip_recognizer.py [hunk 2]
```diff
@@ -50,3 +51,344 @@ def test_when_all_ips_then_succeed(
...
+        ("IP192.168.1.1text", 0, (), (),),
+        ("text192.168.1.1", 0, (), (),),
+        ("(192.168.1.1)", 1, ((1, 12),), ((0.6, 0.81),),),
+        ("'2001:db8::1'", 1, ((1, 12),), ((0.6, 0.81),),),
+        ("IP: 192.168.1.1.", 1, ((4, 15),), ((0.6, 0.81),),),
+        ("192.168.1.1,", 1, ((0, 11),), ((0.6, 0.81),),),
+        ("[2001:db8::1]", 1, ((1, 12),), ((0.6, 0.81),),),
+        ("fe80::1text", 0, (), (),),
+        ("text2001:db8::1", 0, (), (),),
+        # fmt: on
```
[the remark this answers is not in the document]
Comment (kennionblack) [reply presidio-analyzer/tests/test_ip_recognizer.py:310]: Test cases added.

File: presidio-analyzer/tests/test_ip_recognizer.py [hunk 3]
```diff
@@ -50,3 +51,344 @@ def test_when_all_ips_then_succeed(
...
+        # fmt: off
+        # Should be rejected by validation
+        ("256.256.256.256", 0, (), (),),
+        ("192.168.1.256", 0, (), (),),
+        ("gggg:hhhh::1234", 0, (), (),),
+        ("192.168.1", 0, (), (),),
+        ("300.168.1.1", 0, (), (),),
+        ("192.168.1.1.1", 1, ((0, 11),), ((0.6, 0.81),),),
+        ("12345:db8::1", 0, (), (),),
+        ("2001::db8::1", 0, (), (),),
```
[the remark this answers is not in the document]
Comment (kennionblack) [reply presidio-analyzer/tests/test_ip_recognizer.py:345]: I moved the `192.168.1.1.1` case to the boundary cases block and added a comment to explain the behavior.

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kennionblack) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py]: This has been added to the regex.

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py [hunk 4]
```diff
@@ -15,19 +15,29 @@ class IpRecognizer(PatternRecognizer):
...
 
     PATTERNS = [
+        Pattern(
+            "IPv4_mapped",
+            r"(?<![0-9A-Fa-f:])::(?:ffff(?::0{1,4})?:)(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\b",
+            0.6,
+        ),
+        Pattern(
+            "IPv4_embedded",
+            r"(?<![0-9A-Fa-f:])(?:[0-9A-Fa-f]{1,4}:){1,4}:(?:[0-9A-Fa-f]{1,4}:){0,4}(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\b",
```
[the remark this answers is not in the document]
Comment (kennionblack) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py]: I made this consistent with the IPv6 lookbehind.

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py [hunk 5]
```diff
@@ -15,19 +15,29 @@ class IpRecognizer(PatternRecognizer):
     """
 
     PATTERNS = [
+        Pattern(
+            "IPv4_mapped",
+            r"(?<![0-9A-Fa-f:])::(?:ffff(?::0{1,4})?:)(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\b",
```
[the remark this answers is not in the document]
Comment (kennionblack) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py]: I made this consistent with the IPv6 lookbehind.

File: presidio-analyzer/tests/test_ip_recognizer.py [hunk 6]
```diff
@@ -27,6 +27,7 @@ def entities():
         ("684D:1111:222:3333:4444:5555:77", 0, (), (),),
         ("2345:0425:2CA1:0000:0000:0567:5673:23b5", 1, (), (),),  # noqa: E501
         ("2345:0425:2CA1::0567:5673:23b5", 1, (), (),),
+        ("2400:c401::5054:ff:fe1b:b031", 1, (), (),),
         ("Use local ipv6 ::", 1, ((15, 17),), ((0.05, 0.15),),),
```
[the remark this answers is not in the document]
Comment (kennionblack) [reply presidio-analyzer/tests/test_ip_recognizer.py:31]: This is consistent with other tests in this block and represents the IP address reported in the initial issue. Because the whole string should get marked as an IPv6 address, I think the original assumption in the other tests in this block was that it would be redundant to set indices for the beginning and end of the string.

There are more specific tests that implicitly cover this behavior later anyways.

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py [hunk 7]
```diff
@@ -15,19 +15,29 @@ class IpRecognizer(PatternRecognizer):
...
+            r"(?<![\w:])(?:(?:[0-9A-Fa-f]{1,4}:){7}[0-9A-Fa-f]{1,4}|(?:[0-9A-Fa-f]{1,4}:){1,7}:|:(?::[0-9A-Fa-f]{1,4}){1,7}|(?:[0-9A-Fa-f]{1,4}:){1,6}:[0-9A-Fa-f]{1,4}|(?:[0-9A-Fa-f]{1,4}:){1,5}(?::[0-9A-Fa-f]{1,4}){1,2}|(?:[0-9A-Fa-f]{1,4}:){1,4}(?::[0-9A-Fa-f]{1,4}){1,3}|(?:[0-9A-Fa-f]{1,4}:){1,3}(?::[0-9A-Fa-f]{1,4}){1,4}|(?:[0-9A-Fa-f]{1,4}:){1,2}(?::[0-9A-Fa-f]{1,4}){1,5}|[0-9A-Fa-f]{1,4}:(?::[0-9A-Fa-f]{1,4}){1,6}|:(?::[0-9A-Fa-f]{1,4}){1,6})(?![\w:]|\.\d)",
             0.6,
         ),
         Pattern(
-            "IPv6",
-            r"::",
+            "IPv6_unspecified",
+            r"(?<!\w)::(?!\w)",
```
[the remark this answers is not in the document]
Comment (kennionblack) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py]: Good catch, fixed.

File: presidio-analyzer/tests/test_ip_recognizer.py [hunk 8]
```diff
@@ -50,3 +51,353 @@ def test_when_all_ips_then_succeed(
...
             res, entities[0], st_pos, fn_pos, st_score, fn_score
         )
+
+
+@pytest.mark.parametrize(
+    "text, expected_len, expected_positions, expected_score_ranges",
+    [
+        # fmt: off
+        ("2400:c401::5054:ff:fe1b:b031", 1, ((0, 28),), ((0.6, 0.81),),),
+        ("fe80::1", 1, ((0, 7),), ((0.6, 0.81),),),
```
Comment (omri374) [inline presidio-analyzer/tests/test_ip_recognizer.py:61]: not an expert in IP addresses, but isn't this a valid ip address too? `fe80::1%eth0`? Will the new patterns detect it?
How about this one? `fe80::a00:27ff:fe4e:66a1%en0`
Among other RFC-compliant scoped IPv6 addresses

Comment (kennionblack) [reply presidio-analyzer/tests/test_ip_recognizer.py:61]: These patterns will detect up to the % character but not after. I added an optional block to the IPv6 regex that handles the interface logic.

I also did some digging on IP RFCs and found RFC 4291 2.5.5.1 considers `::192.168.0.1` to be a valid address even though the format should be deprecated by now, so I modified the `IPv4_mapped` pattern to account for this style.

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py [hunk 9]
```diff
@@ -60,6 +70,6 @@ def invalidate_result(self, pattern_text: str) -> bool:
         :return: True if invalidated
         """
         try:
-            ipaddress.ip_address(pattern_text)
+            ipaddress.ip_interface(pattern_text)
         except ValueError:
```
[the remark this answers is not in the document]
Comment (kennionblack) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py:74]: `ipaddress.ip_interface()` does accept the `%zone` suffix, which is evidenced by the tests on lines 65, 427, and 428 passing.

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py [hunk 10]
```diff
@@ -15,19 +15,29 @@ class IpRecognizer(PatternRecognizer):
     """
 
     PATTERNS = [
+        Pattern(
+            "IPv4_mapped",
```
[the remark this answers is not in the document]
Comment (kennionblack) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/generic/ip_recognizer.py:19]: This IPv4 compatible pattern without the `ffff` segment is a deprecated format per [this link](https://datatracker.ietf.org/doc/html/rfc4291#section-2.5.5.1). As such, I don't see much reason to change the name to indicate compatibility with a format that should be rare in actual text.

**Comment by kennionblack** [comment]
Notable changes from the original PR description are below:

- Modified `IPv4_mapped` pattern to also recognize IPv4-compatible IPv6 addresses ([source](https://datatracker.ietf.org/doc/html/rfc4291#section-2.5.5.1)). This format is deprecated and unlikely to show up all that often, but the regex is now capable of recognizing this format.
-  Added an optional capturing group for zone suffixes to the IPv6 regex patterns (e.g. `fe80::1%eth0`). [RFC 4007](https://datatracker.ietf.org/doc/html/rfc4007#section-11.2) is where the idea of a zone suffix was introduced, meaning that these suffixes are not valid parts of IPv4 addresses. I tested adding a zone suffix to an IPv4 pattern and `ipaddress.ip_interface()` rejected the pattern, which would appear to support the idea that these suffixes are not valid for IPv4.
-  Added capturing groups for CIDR suffixes (e.g. `192.168.0.1/24` and `fe::1/128`), which are commonly used in networking contexts to indicate which subnet or interface is in use. 
- Switched the method call in `invalidate_result` from `ipaddress.ip_address()` to `ipaddress.ip_interface()` to allow addresses with CIDR suffixes to not throw a ValueError.

**Comment by omri374** [comment]
Thanks! Great work

**Review by omri374 — APPROVED** [review]

**Review by SharonHart — APPROVED** [review]
[kennionblack](https://github.com/kennionblack) Amazing, thanks for both contributions!
